File: solutions/java/phone-number/1/src/main/java/PhoneNumber.java
import java.util.stream.Collectors;

class PhoneNumber {
    private String value;

    PhoneNumber(String userInput) {

        boolean hasLetters = userInput.chars().mapToObj(i -> (char) i).anyMatch(c -> Character.isAlphabetic(c));

        if (hasLetters)
            throw new IllegalArgumentException("letters not permitted");

        boolean hasUnexpectedPunctuation = userInput.chars().mapToObj(i -> (char) i)
                .anyMatch(c -> c == ',' || c == ':' || c == ';');

        if (hasUnexpectedPunctuation)
            throw new IllegalArgumentException("punctuations not permitted");

        String cleansedNumber = userInput.chars().mapToObj(i -> (char) i).filter(i -> Character.isDigit(i))
                .map(c -> c.toString()).collect(Collectors.joining());

        if (cleansedNumber.length() > 11) {
            throw new IllegalArgumentException("more than 11 digits");
        }

        if (cleansedNumber.length() == 11) {
            if (cleansedNumber.charAt(0) != '1') {
                throw new IllegalArgumentException("11 digits must start with 1");
            } else {
                // Strip leading 1
                cleansedNumber = cleansedNumber.substring(1);
            }
        }

        if (cleansedNumber.length() != 10)
            throw new IllegalArgumentException("incorrect number of digits");

        if (cleansedNumber.charAt(0) == '0')
            throw new IllegalArgumentException("area code cannot start with zero");

        if (cleansedNumber.charAt(0) == '1')
            throw new IllegalArgumentException("area code cannot start with one");

        if (cleansedNumber.charAt(3) == '0')
            throw new IllegalArgumentException("exchange code cannot start with zero");

        if (cleansedNumber.charAt(3) == '1')
            throw new IllegalArgumentException("exchange code cannot start with one");

        this.value = cleansedNumber;
    }

    String getNumber() {
        return this.value;
    }
}

File: solutions/javascript/phone-number/1/phone-number.js
export const clean = (input) => {
  // Validate
  if (/[a-zA-Z]/.test(input)) throw new Error("Letters not permitted");
  if (/[^0-9\(\)\- .+]/.test(input))
    throw new Error("Punctuations not permitted");

  // Normalize
  let digits = input
    .split("")
    .filter((c) => /\d/.test(c))
    .join("");

  // Validate Length
  if (digits.length > 11) throw new Error("More than 11 digits");
  if (digits.length <= 9) throw new Error("Incorrect number of digits");
  if (digits.length == 11) {
    if (digits.charAt(0) != 1) {
      throw new Error("11 digits must start with 1");
    } else {
      digits = digits.substring(1);
    }
  }

  // Validate leading area and exchange code digits
  if (digits.length == 10) {
    if (digits.charAt(0) == 0) {
      throw new Error("Area code cannot start with zero");
    }
    if (digits.charAt(0) == 1) {
      throw new Error("Area code cannot start with one");
    }
    if (digits.charAt(3) == 0) {
      throw new Error("Exchange code cannot start with zero");
    }
    if (digits.charAt(3) == 1) {
      throw new Error("Exchange code cannot start with one");
    }
  }

  return digits;
};
